Add Local Fonts API article (#3615)
GoogleChrome/web.dev: src/site/content/en/blog/local-fonts/index.md

---
layout: post

> title: Create advanced typography content with local fonts

Author: ```suggestion
title: Use advanced typography with local fonts
```

> title: Create advanced typography content with local fonts

Author: That's a good suggestion. Thanks!

Company fonts apart, even just the list of installed fonts can be identifying.
The situation with this attack vector has become so bad that recently the WebKit team [decided](https://webkit.org/tracking-prevention/#table-of-contents-toggle:~:text=Changed%20font%20availability%20to%20web%20content,but%20not%20locally%20user%2Dinstalled%20fonts) to *"only include [in the list available fonts] web fonts and fonts that come with the operating system, but not locally user-installed fonts"*. (And here I am, with an article on granting access to local fonts.)

## The Local Font Access API

The beginning of this article may have put you in a negative mood. Can we really not have nice things? Fret not. We think we can, and maybe [everything is not hopeless](http://hyperboleandahalf.blogspot.com/2013/05/depression-part-two.html#Blog1:~:text=like-,hopeless). But first, let me answer a question that you might be asking yourself.

### Why do we need the Local Font Access API when there are web fonts?

> Professional-quality design and graphics tools have historically been difficult to deliver on the web. One stumbling block has been an inability to access and use the full variety of professionally constructed and hinted fonts that designers have locally installed. Web fonts enable some publishing use-cases, but fail to enable programmatic access to the vector glyph shapes and a set other font internals, and the font binary data.

Author: It's indeed meant as "access to (the vector glyph shapes and a set other font internals), and (the font binary data)".

> Professional-quality design and graphics tools have historically been difficult to deliver on the web. One stumbling block has been an inability to access and use the full variety of professionally constructed and hinted fonts that designers have locally installed. Web fonts enable some publishing use-cases, but fail to enable programmatic access to the vector glyph shapes and a set other font internals, and the font binary data.

Author: ```suggestion
Professional-quality design and graphics tools have historically been difficult to deliver on the web. One stumbling block has been an inability to access and use the full variety of professionally constructed and hinted fonts that designers have locally installed. Web fonts enable some publishing use-cases, but fail to enable programmatic access to the vector glyph shapes and a set of other font internals, and the font binary data.
```

> Professional-quality design and graphics tools have historically been difficult to deliver on the web. One stumbling block has been an inability to access and use the full variety of professionally constructed and hinted fonts that designers have locally installed. Web fonts enable some publishing use-cases, but fail to enable programmatic access to the vector glyph shapes and a set other font internals, and the font binary data.

Reviewer: We need make this less confusing. What does 'a set of other font internals' mean? Is that the set of other font internals for the vector glyphs?

> Professional-quality design and graphics tools have historically been difficult to deliver on the web. One stumbling block has been an inability to access and use the full variety of professionally constructed and hinted fonts that designers have locally installed. Web fonts enable some publishing use-cases, but fail to enable programmatic access to the vector glyph shapes and a set other font internals, and the font binary data.

Author: I gave it [another shot](https://github.com/GoogleChrome/web.dev/pull/3615/commits/c2e6c5fd51b7aa3c4ae7e5d4ca3aafe6a98b4d8b).

[source code](https://glitch.com/edit/#!/local-font-access?path=README.md%3A1%3A0).
The demo showcases a custom element `<font-select>` that implements a local font picker.

<div class="glitch-embed-wrap" style="height: 500px; width: 100%;">
  <iframe
    src="https://glitch.com/embed/#!/embed/local-font-access?path=index.html&previewSize=100"
    title="local-font-access on Glitch"
    allow="local-fonts"
    style="height: 100%; width: 100%; border: 0;">
  </iframe>
</div>

## Privacy considerations

The `"local-fonts"` permission appears to provide a highly fingerprintable surface.
However, browsers are free to return anything they like.
For example, anonymity-focused browsers may choose to only provide
a set of default fonts built into the browser.
Similarly, browsers are not required to provide table data exactly as it appears on disk.

Wherever possible, the Local Font Access API is designed to only expose exactly the information
needed to enable the mentioned use cases.
System APIs may produce a list of installed fonts not in a random or a sorted order,
but in the order of font installation.
Returning exactly the list of installed fonts given by such a system API

> can expose additional entropy bits, and use cases we want to enable

Author: ```suggestion
can expose additional data that may be used for fingerprinting, and use cases we want to enable
```

> can expose additional entropy bits, and use cases we want to enable

Author: Changed. Thanks!

are not assisted by retaining this ordering.
As a result, this API requires that the returned data be sorted before being returned.

## Security and permissions

The Chrome team has designed and implemented the Local Font Access API using the core principles defined in [Controlling Access to Powerful Web Platform Features][powerful-apis], including user control, transparency, and ergonomics.

### User control

Access to a user's fonts is fully under their control and will not be allowed unless the `"local-fonts"` permission, as listed in the
[permission registry](https://w3c.github.io/permissions/#permission-registry), is granted.

### Transparency

> The fact whether a site has been granted access to the user's local fonts will be visible in the *site information* sheet.

Reviewer: ```suggestion
Whether a site has been granted access to the user's local fonts will be visible in the *site information* sheet.
```

> The fact whether a site has been granted access to the user's local fonts will be visible in the *site information* sheet.

Reviewer: What's a site information sheet? Maybe add a link.

Reviewer: Is that the thing that pops up when you click the lock icon in the address bar? Maybe show a picture instead.

Author: It is, yes. For clarity, I have added a link: https://support.google.com/chrome/answer/114662?hl=en&co=GENIE.Platform=Desktop.

> Whether a site has been granted access to the user's local fonts will be visible in the *site information* sheet.

Author: ```suggestion
Whether a site has been granted access to the user's local fonts will be visible in the
[site information sheet](https://support.google.com/chrome/answer/114662?hl=en&co=GENIE.Platform=Desktop).
```

---
layout: post
title: Use advanced typography with local fonts
subhead: >
  Learn how the Local Font Access API allows you to access the user's locally installed fonts and obtain low-level details about them.
tags:
  - blog
  - fonts
  - capabilities
authors:
  - thomassteiner
description: >
  The Local Fonts API enumerates the user's installed local
  fonts and provides low-level access to the various
  TrueType/OpenType tables.

> date: 2020-08-03

Author: ```suggestion
date: 2020-08-12
```

Company fonts apart, even just the list of installed fonts can be identifying.
The situation with this attack vector has become so bad that recently the WebKit team [decided](https://webkit.org/tracking-prevention/#table-of-contents-toggle:~:text=Changed%20font%20availability%20to%20web%20content,but%20not%20locally%20user%2Dinstalled%20fonts) to *"only include [in the list available fonts] web fonts and fonts that come with the operating system, but not locally user-installed fonts"*. (And here I am, with an article on granting access to local fonts.)

## The Local Font Access API

The beginning of this article may have put you in a negative mood. Can we really not have nice things? Fret not. We think we can, and maybe [everything is not hopeless](http://hyperboleandahalf.blogspot.com/2013/05/depression-part-two.html#Blog1:~:text=like-,hopeless). But first, let me answer a question that you might be asking yourself.

### Why do we need the Local Font Access API when there are web fonts?

> Professional-quality design and graphics tools have historically been difficult to deliver on the web. One stumbling block has been an inability to access and use the full variety of professionally constructed and hinted fonts that designers have locally installed. Web fonts enable some publishing use-cases, but fail to enable programmatic access to the vector glyph shapes and a set of other font internals, and the font binary data.

Author: ```suggestion
Professional-quality design and graphics tools have historically been difficult to deliver on the web.
One stumbling block has been an inability to access and use the full variety of
professionally constructed and hinted fonts that designers have locally installed.
Web fonts enable some publishing use-cases, but fail to enable programmatic access to
the vector glyph shapes and font tables used by rasterizers to render the glyph outlines.
There is likewise no way to access a web font's binary data.
```

---
layout: post
title: Use advanced typography with local fonts
subhead: >
  Learn how the Local Font Access API allows you to access the user's locally installed fonts and obtain low-level details about them.
tags:
  - blog
  - fonts
  - capabilities
authors:
  - thomassteiner
description: >
  The Local Fonts API enumerates the user's installed local
  fonts and provides low-level access to the various
  TrueType/OpenType tables.

> date: 2020-08-12

Author: ```suggestion
date: 2020-08-18
```

> date: 2020-08-18

Author: ```suggestion
date: 2020-08-19
```

File a spec issue on the corresponding [GitHub repo][issues], or add your thoughts to an existing issue.

### Report a problem with the implementation

Did you find a bug with Chrome's implementation? Or is the implementation different from the spec?
File a bug at [new.crbug.com](https://new.crbug.com). Be sure to include as much detail as you can, simple instructions for reproducing, and enter `Blink>Fonts` in the **Components** box. [Glitch](https://glitch.com/) works great for sharing quick and easy repros.

### Show support for the API

Are you planning to use the Local Font Access API? Your public support helps the Chrome team to prioritize features and shows other browser vendors how critical it is to support them.

Send a Tweet to [@ChromiumDev][cr-dev-twitter] with the `#LocalFontAccess` hashtag and let us know where and how you're using it.

## Helpful links

- [Explainer][explainer]
- [Spec draft][spec]
- [Chromium bug for font enumeration][cr-bug-enum]
- [Chromium bug for font table access][cr-bug-table]
- [ChromeStatus entry][cr-status]
- [GitHub repo][issues]

> - [Tag review][tag-review]

Author: ```suggestion
- [Tag review][tag-review]
- [Mozilla standards position][mozilla]
```

## Acknowledgements

The Local Font Access API spec was edited by
[Emil A. Eklund](https://www.linkedin.com/in/emilaeklund/),
[Alex Russell](https://infrequently.org/),
[Joshua Bell](https://www.linkedin.com/in/joshuaseanbell/), and
[Olivier Yiptong](https://github.com/oyiptong/).
This article was reviewed by
[Joe Medley](https://github.com/jpmedley),
[Dominik Röttsches](https://fi.linkedin.com/in/dominik-r%C3%B6ttsches-7323684), and
[Olivier Yiptong](https://github.com/oyiptong/).
Hero image by [Brett Jordan](https://unsplash.com/@brett_jordan) on [Unsplash](https://unsplash.com/photos/qrjvkj-oS-M).

[tag-review]: https://github.com/w3ctag/design-reviews/issues/399
[explainer]: https://github.com/WICG/local-font-access
[spec]: https://wicg.github.io/local-font-access/
[powerful-apis]: https://chromium.googlesource.com/chromium/src/+/lkgr/docs/security/permissions-for-powerful-web-platform-features.md
[issues]: https://github.com/WICG/local-font-access/issues
[cr-status]: https://chromestatus.com/feature/6234451761692672
[cr-bug-enum]: https://bugs.chromium.org/p/chromium/issues/detail?id=535764
[cr-bug-table]: https://crbug.com/982054

> [cr-dev-twitter]: https://twitter.com/ChromiumDev

Author: ```suggestion
[cr-dev-twitter]: https://twitter.com/ChromiumDev
[mozilla]: https://github.com/mozilla/standards-positions/issues/401
```
